ra-tls/libenclave-tls/enclave-tls.c

```diff
@@ -0,0 +1,262 @@
+		} while (0)
+
+		DLSYM(wolfSSL_free);
+		DLSYM(wolfSSL_CTX_set_verify);
+		DLSYM(create_key_and_x509);
+		DLSYM(wolfSSL_read);
+		DLSYM(wolfSSL_write);
+		DLSYM(wolfSSL_CTX_new);
+		DLSYM(printf_identity);
+		DLSYM(wolfTLSv1_2_client_method);
+		DLSYM(wolfSSL_X509_get_der);
+		DLSYM(wolfSSL_get_peer_certificate);
+		DLSYM(wolfSSL_connect);
+		DLSYM(wolfSSL_negotiate);
+		DLSYM(wolfSSL_set_fd);
+		DLSYM(wolfSSL_new);
+		DLSYM(wolfTLSv1_2_server_method);
+		DLSYM(wolfSSL_CTX_free);
+		DLSYM(wolfSSL_Cleanup);
+		DLSYM(wolfSSL_Debugging_ON);
+		DLSYM(wolfSSL_Debugging_OFF);
+		DLSYM(wolfSSL_Init);
+#undef DLSYM
+	} else {
+#define DLSYM(fn)                                                                       \
+                do {                                                                    \
+                        ftab.fn = &fn;                                                  \
+                } while (0)
+
+		DLSYM(wolfSSL_free);
+		DLSYM(wolfSSL_CTX_set_verify);
+		DLSYM(create_key_and_x509);
+		DLSYM(wolfSSL_read);
+		DLSYM(wolfSSL_write);
+		DLSYM(wolfSSL_CTX_new);
+		DLSYM(printf_identity);
+		DLSYM(wolfTLSv1_2_client_method);
+		DLSYM(wolfSSL_X509_get_der);
+		DLSYM(wolfSSL_get_peer_certificate);
+		DLSYM(wolfSSL_connect);
+		DLSYM(wolfSSL_negotiate);
+		DLSYM(wolfSSL_set_fd);
+		DLSYM(wolfSSL_new);
+		DLSYM(wolfTLSv1_2_server_method);
+		DLSYM(wolfSSL_CTX_free);
+		DLSYM(wolfSSL_Cleanup);
+		DLSYM(wolfSSL_Debugging_ON);
+		DLSYM(wolfSSL_Debugging_OFF);
+		DLSYM(wolfSSL_Init);
+#undef DLSYM
+	}
+
+	ftab.wolfSSL_Debugging_ON();
+
+	ftab.wolfSSL_Init(ctx->quote_type);
+
+	WOLFSSL_CTX *wolf_ctx;
+	if (ctx->is_server)
+		wolf_ctx = ftab.wolfSSL_CTX_new(ftab.wolfTLSv1_2_server_method());
+	else
+		wolf_ctx = ftab.wolfSSL_CTX_new(ftab.wolfTLSv1_2_client_method());
+	if (!wolf_ctx) {
+		fprintf(stderr, "ERROR: failed to create WOLFSSL_CTX\n");
+		ftab.wolfSSL_Cleanup();
+	}
+
+	if (ctx->is_server) {
+		ftab.create_key_and_x509(wolf_ctx);
+		if (ctx->mutual) {
+			ftab.wolfSSL_CTX_set_verify(wolf_ctx,
+						    WOLFSSL_VERIFY_PEER |
+						    WOLFSSL_VERIFY_FAIL_IF_NO_PEER_CERT,
+						    cert_verify_callback);
+		}
+	} else {
+		if (ctx->mutual) {
+			ftab.create_key_and_x509(wolf_ctx);
+		}
+		ftab.wolfSSL_CTX_set_verify(wolf_ctx, SSL_VERIFY_PEER,
+					    cert_verify_callback);
+	}
+
+	ctx->ssl = ftab.wolfSSL_new(wolf_ctx);
+	if (!ctx->ssl) {
+		fprintf(stderr, "ERROR: failed to create WOLFSSL object\n");
+		ftab.wolfSSL_CTX_free(ctx->ssl);
+	}
+
+	/* Attach wolfSSL to the socket */
+	ftab.wolfSSL_set_fd(ctx->ssl, ctx->sockfd);
+}
+
+int enclave_tls_connect(enclave_tls_ctx *ctx)
+{
+	if (ctx->is_server) {
+		fprintf(stderr,
+			"ERROR: can NOT use enclave_tls_connect function in server\n");
+		ftab.wolfSSL_free(ctx->ssl);
+	}
+
+	if (ftab.wolfSSL_connect(ctx->ssl) != SSL_SUCCESS) {
+		ftab.wolfSSL_free(ctx->ssl);
+	}
+
+	WOLFSSL_X509 *srvcrt = ftab.wolfSSL_get_peer_certificate(ctx->ssl);
```

Comment:
Add a check to the validation of ctx and ctx->ssl. Because anyone can randomly call any enclave tls api you provide. If init api is not called, the null pointer dereference will be triggered.